### PR TITLE
User permission management with Pundit

### PR DESCRIPTION
**Pull request by joaquinvalentin**
- Added Pundit
- Generated a UserPolicy to restrict CRUD in Users to admin type and included in the methods
- Added new errors due to new cases (or cases that I didn't contemplate)
- Refactored some specs for admin cases and create others for non-admin.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -78,13 +78,15 @@ gem 'jwt', '~> 2.3'
 gem 'blueprinter', '~> 0.25.3'
 
 group :test do
-  gem 'database_cleaner-active_record'
+  gem 'database_cleaner-active_record', '~> 2.0.1'
```
Comment (guillezorron) [inline Gemfile:81]: 👏

File: app/controllers/api/v1/users_controller.rb [hunk 2]
```diff
@@ -1,34 +1,42 @@
...
   # POST /users
   def create
     user = User.new(user_params)
+    authorize user
     if user.save
       render json: UserSerializer.render(user), status: :created
-    else
-      render_error(4022, user.errors.messages[:error])
+    elsif email_taken
+      render_error(4010)
```
Comment (guillezorron) [inline app/controllers/api/v1/users_controller.rb]: I prefer to use it at the beginning of this method, like `return render_error(4010) if user_by_email` in order to stop running code if there is an error.

File: app/controllers/api/v1/users_controller.rb [hunk 3]
```diff
@@ -59,7 +73,7 @@ def user
     @user ||= User.find(params[:id])
   end
 
-  def check_permissions
-    render_error(4011) unless current_user&.admin? || current_user&.id == user.id
+  def email_taken
+    !User.find_by_email(user_params[:email]).nil?
```
Comment (guillezorron) [inline app/controllers/api/v1/users_controller.rb]:
```suggestion
  def user_by_email
    User.find_by_email(user_params[:email])
```

File: app/controllers/api/v1/users_controller.rb [hunk 4]
```diff
@@ -1,34 +1,42 @@
...
   end
 
   # PATCH/PUT /users/1
   def update
+    authorize user
     if user.update(user_params)
       render json: UserSerializer.render(user), status: :ok
-    else
-      render_error(4023, user.errors.messages[:error])
+    elsif email_taken
```
Comment (guillezorron) [inline app/controllers/api/v1/users_controller.rb]: Same comment as before

File: app/controllers/api/v1/users_controller.rb [hunk 5]
```diff
@@ -48,8 +56,14 @@ def destroy
     render_error(4011)
   end
 
+  rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
+
   private
 
+  def user_not_authorized(exception)
+    render_error(4011, exception.message)
```
Comment (guillezorron) [inline app/controllers/api/v1/users_controller.rb]: This error is not in errors.yml

File: config/errors.yaml [hunk 6]
```diff
@@ -40,3 +40,11 @@
   :http_status: 401
   :error_code: 4111
   :description: 'Product can not be deleted or updated due to being unauthorized'
+4009:
```
Comment (guillezorron) [inline config/errors.yaml]: Can we maintain the order of error codes ascendant for this file?

File: config/errors.yaml [hunk 7]
```diff
@@ -40,3 +40,11 @@
...
+  :http_status: 401
+  :error_code: 4009
+  :description: 'User is not logged in'
```
Comment (guillezorron) [inline config/errors.yaml]:
```suggestion
  :description: 'Unauthorized request'
```

File: app/controllers/concerns/error_handler.rb [hunk 8]
```diff
@@ -6,4 +6,7 @@ def render_error(error_code, details = {})
       Error.instance.error_for(error_code, details)
     ), status: Error.instance.http_status_for(error_code)
   end
+
+  class AuthenticationError < StandardError
```
Comment (guillezorron) [inline app/controllers/concerns/error_handler.rb]: Why are you defining this here?

File: config/errors.yaml [hunk 9]
```diff
@@ -40,3 +40,11 @@
...
+4010:
+  :http_status: 422
+  :error_code: 4010
+  :description: 'The email is already in use'
```
Comment (guillezorron) [inline config/errors.yaml]: Missing end of line at the end of file

File: app/controllers/concerns/error_handler.rb (hunk 8 above)
Comment (joaquinvalentin) [reply app/controllers/concerns/error_handler.rb]: This is for cases when the JWT is invalid (this is not new, but was moved to errorHandler module).

File: app/controllers/api/v1/users_controller.rb [hunk 10]
```diff
@@ -29,6 +36,7 @@ def update
 
   # DELETE /users/1
   def destroy
+    authorize user
```
Comment (MarcosFierro) [inline app/controllers/api/v1/users_controller.rb]: why not do this as a before action?

File: spec/controllers/api/v1/users_controller_spec.rb [hunk 11]
```diff
@@ -54,38 +106,88 @@ def create_call(user_params)
...
 
       it 'returns the error code 4022' do
+        authenticate_user(admin)
         create_call(new_user)
         expect(JSON.parse(response.body)['error_code']).to be(4022)
       end
 
       it 'returns code 422' do
+        authenticate_user(admin)
         create_call(new_user)
```
Comment (MarcosFierro) [inline spec/controllers/api/v1/users_controller_spec.rb]: you could define this as a setup method. Also check if admin is not becoming a mistery guest (method is way too up on the file doesn't it?)

File: spec/support/authentication_helper.rb [hunk 12]
```diff
@@ -14,4 +14,12 @@ def authenticate_user(user)
     jwt = JsonWebToken.encode(payload)
     @request.headers['Authorization'] = "Bearer #{jwt}"
   end
+
+  def create_authenticated_admin
+    user = User.first || FactoryBot.create(:user, :is_admin)
```
Comment (MarcosFierro) [inline spec/support/authentication_helper.rb]: can you be sure that the first User is going to be admin?

File: app/controllers/concerns/error_handler.rb (hunk 8 above)
Comment (guillezorron) [reply app/controllers/concerns/error_handler.rb]: I think that this class error defined makes more sense in the file that is using it.

Comment (guillezorron) [reply app/controllers/concerns/error_handler.rb]: Please delete it.

File: app/controllers/api/v1/users_controller.rb [hunk 13]
```diff
@@ -11,20 +12,21 @@ def show
...
-      render json: UserSerializer.render(user), status: :created
-    else
-      render_error(4104, user.errors.messages[:error])
-    end
+    authorize user
+    return render_error(4106) if user_by_email
+
+    return render json: UserSerializer.render(user), status: :created if user.save
+
+    render_error(4104, user.errors.messages[:error])
```
Comment (guillezorron) [inline app/controllers/api/v1/users_controller.rb:20]: Awesome 😁

**Review by guillezorron — APPROVED** [review]

**Pull request merged**